File: Webpage/server/webpage.server/webpage.server/src/main/java/abbruzzese/webpage/webpage/server/webpage/Service/MagazineService.java
package abbruzzese.webpage.webpage.server.webpage.Service;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import abbruzzese.webpage.webpage.server.webpage.DTO.DTOMagazine;
import abbruzzese.webpage.webpage.server.webpage.Mappers.MagazineMapper;
import abbruzzese.webpage.webpage.server.webpage.Repository.MagazineRepository;
import abbruzzese.webpage.webpage.server.webpage.entities.Magazine;
import jakarta.transaction.Transactional;

@Service
public class MagazineService {

	@Autowired
	private MagazineMapper MagazineMapper;

    @Autowired
    private MagazineRepository magazineRepository;

    @Transactional
    public List<DTOMagazine> getMagazinesByYearRange(String startYear, String endYear) {
        List<Magazine> magazines = magazineRepository.findRevistasByAnioRange(startYear, endYear);
        
        return magazines.stream()
                        .map(MagazineMapper::MagazinetoDTO)
                        .collect(Collectors.toList());
    }

	@Transactional
    public DTOMagazine saveMagazine(DTOMagazine dto) {
        Magazine magazine = MagazineMapper.DTOtoMagazine(dto);
        magazine = magazineRepository.save(magazine);
        return MagazineMapper.MagazinetoDTO(magazine);
    }

	@Transactional
    public List<DTOMagazine> getAllMagazines() {
        return ((List<Magazine>) magazineRepository.findAll())
				.stream()
                .map(MagazineMapper::MagazinetoDTO)
                .collect(Collectors.toList());
    }

	@Transactional
    public Optional<DTOMagazine> getMagazineByName(String nombre) {
        return magazineRepository.findByNombre(nombre)
                .map(MagazineMapper::MagazinetoDTO);
    }

	@Transactional
	public DTOMagazine updateMagazine(String id, DTOMagazine dto) {
        Optional<Magazine> existingMagazineOpt = magazineRepository.findById(id);
        if (existingMagazineOpt.isPresent()) {
            Magazine existingMagazine = existingMagazineOpt.get();
            existingMagazine.setNombre(dto.getNombre());
            existingMagazine.setDescripcion(dto.getDescripcion());
            existingMagazine.setAño(dto.getAño());
            existingMagazine.setImageString(dto.getImageString());
            magazineRepository.save(existingMagazine); // Guarda los cambios
            return MagazineMapper.MagazinetoDTO(existingMagazine);
        }
        return null;
    }

    public void deleteMagazine(String id) {
		magazineRepository.deleteById(id);
    }

    // Método para obtener el enlace de la última revista
    @Transactional
    public String obtenerUltimoEnlace() {
        Optional<Magazine> ultimaRevistaOpt = magazineRepository.findLastMagazine();
        return ultimaRevistaOpt.map(Magazine::getEnlace).orElse(null); // Retorna el enlace o null si no se encuentra
    }

    @Transactional
    public Map<String, Object> obtenerUltimoLink() {
    Optional<Magazine> ultimaRevistaOpt = magazineRepository.findLastMagazine();
    Map<String, Object> response = new HashMap<>();

    if (ultimaRevistaOpt.isPresent()) {
        Magazine ultimaRevista = ultimaRevistaOpt.get();
        response.put("enlace", ultimaRevista.getEnlace());
        response.put("Nombre", ultimaRevista.getNombre());
        response.put("ImageString", ultimaRevista.getImageString());
    }

    return response; // Retorna el Map con los 3 atributos
    }

}

File: PHP_API/SERVICE/MagazineService.php
<?php

namespace App\Service;

use App\Entity\Magazine;
use App\Entity\Newsletter;
use App\Repository\MagazineRepository;
use Doctrine\ORM\EntityManagerInterface;
use App\Mapper\MagazineMapper;
require_once __DIR__ . '/../sendMagazineEmail.php';

class MagazineService{

    private $magazineRepository;
    private EntityManagerInterface $entityManager;


    public function __construct(MagazineRepository $magazineRepository, EntityManagerInterface $entityManager)
    {
        $this->magazineRepository = $magazineRepository;
        $this->entityManager = $entityManager;
    }

    public function getMagazineByNombre($nombre)
    {

        $magazine = $this->magazineRepository->findByNombre($nombre);

        if ($magazine) {
            return  MagazineMapper::MagazineToDTO($magazine);
        }
    }

    public function getAllMagazine()
    {
        $magazine = $this->magazineRepository->findAll();  // Método findAll() de Doctrine
        return array_map([MagazineMapper::class, 'MagazinetoDTO'], $magazine);  // Convertimos todos los usuarios en DTOs
    }

    public function deleteMagazine($nombre)
    {
        $magazine = $this->magazineRepository->findByNombre($nombre);  // Buscamos al usuario por email
        if ($magazine) {
            $imagestring=$magazine->getimagestring();
            $uploadDir = __DIR__ . '/magazineupload/';  
            $imagePath = $uploadDir . $imagestring;  
    
            // Verifica si la imagen existe y elimínala
            if ($imagestring && file_exists($imagePath)) {
                unlink($imagePath); // Elimina el archivo físico
            }
            try {
                // Verifica si la imagen existe y elimínala
                if ($imagestring && file_exists($imagestring)) {
                    unlink($imagestring); // Elimina el archivo físico
                }
            } catch (\Throwable $e) {
                return [            
                    'message' => 'No se pudo eliminar la imagen: ' . $e->getMessage(),
                ];
            }
       
            $this->entityManager->remove($magazine);  // Usamos remove() de Doctrine
            $this->entityManager->flush();  // Confirmamos los cambios
            return true;
        }
        return false;
    }

        public function handleMagazineUpload($nombre, $descripcion, $año, $imagestring, $enlace)
        {
            // Definir directorio de almacenamiento
            $uploadDir = __DIR__ . '/magazineupload';
            if (!is_dir($uploadDir)) {
                mkdir($uploadDir, 0777, true);
            }
    
            // Generar un nombre único para la imagen
            $fileName = uniqid() . '_' . basename($imagestring['name']);
            $filePath = $uploadDir . DIRECTORY_SEPARATOR . $fileName;
    
            // Mover el archivo al directorio de destino
            if (!move_uploaded_file($imagestring['tmp_name'], $filePath)) {
                return ['error' => 'Error al subir la imagen.'];
            }
    
            // Guardar en la base de datos
            $magazine = new Magazine($nombre,$descripcion,$año,$fileName,$enlace);
    
            $this->entityManager->persist($magazine);
            $this->entityManager->flush();

              // Obtener todos los emails de los suscriptores
              $newsletterRepository = $this->entityManager->getRepository(Newsletter::class);
              $newsletterService = new NewsletterService($newsletterRepository, $this->entityManager);
              $newsletterDTOs = $newsletterService->getAllNewsletter();
              $emails = array_map(fn($dto) => $dto->getEmail(), $newsletterDTOs);

              // Enviar email masivo con Batch Sending
              enviarCorreoRevistaPorLotes($emails, $magazine);
          
            return [
                'success' => true,
                'message' => 'Revista subida correctamente.',
                'magazine' => MagazineMapper::MagazineToDTO($magazine)
            ];
        }
        
    public function getMagazineByEmail($email){

        $magazine = $this->magazineRepository->findByEmail($email);

        if ($magazine) {

            return MagazineMapper::MagazineToDTO($magazine);
            
        }
    }

    public function getUltimoEnlaceMagazine(){

        $ultimoenlacemagazine= $this->magazineRepository->findLastMagazine();

        if ($ultimoenlacemagazine) {
        return $ultimoenlacemagazine ? $ultimoenlacemagazine->getEnlace(): null;
        }

    }

    // Método para obtener los datos de la última revista
    public function obtenerUltimoLink() {
    // Llamar al repositorio para obtener la última revista
    $ultimaRevista = $this->magazineRepository->findLastMagazine();

    // Crear un array asociativo para almacenar los datos
    $response = [];

    // Si se encuentra la revista, agregar los valores al array
    if ($ultimaRevista) {
        $response['enlace'] = $ultimaRevista->getEnlace();
        $response['Nombre'] = $ultimaRevista->getNombre();
        $response['ImageString'] = $ultimaRevista->getImageString();
    }

    // Retornar el array con los 3 atributos
    return $response;
    }

    public function getMagazinesByYearRange(string $startYear, string $endYear): array
{
    $magazines = $this->magazineRepository->findRevistasByAnioRange($startYear, $endYear);
    
    return array_map(fn($magazine) => MagazineMapper::MagazineToDTO($magazine), $magazines);
}

}

?>